elife/config/etc-php-7.0-fpm-pool.d-www.conf

```diff
@@ -0,0 +1,415 @@
+; must be separated by a comma. If this value is left blank, connections will be
+; accepted from any ip address.
+; Default Value: any
+;listen.allowed_clients = 127.0.0.1
+
+; Specify the nice(2) priority to apply to the pool processes (only if set)
+; The value can vary from -19 (highest priority) to 20 (lower priority)
+; Note: - It will only work if the FPM master process is launched as root
+;       - The pool processes will inherit the master process priority
+;         unless it specified otherwise
+; Default Value: no set
+; process.priority = -19
+
+; Choose how the process manager will control the number of child processes.
+; Possible Values:
+;   static  - a fixed number (pm.max_children) of child processes;
+;   dynamic - the number of child processes are set dynamically based on the
+;             following directives. With this process management, there will be
+;             always at least 1 children.
+;             pm.max_children      - the maximum number of children that can
+;                                    be alive at the same time.
+;             pm.start_servers     - the number of children created on startup.
+;             pm.min_spare_servers - the minimum number of children in 'idle'
+;                                    state (waiting to process). If the number
+;                                    of 'idle' processes is less than this
+;                                    number then some children will be created.
+;             pm.max_spare_servers - the maximum number of children in 'idle'
+;                                    state (waiting to process). If the number
+;                                    of 'idle' processes is greater than this
+;                                    number then some children will be killed.
+;  ondemand - no children are created at startup. Children will be forked when
+;             new requests will connect. The following parameter are used:
+;             pm.max_children           - the maximum number of children that
+;                                         can be alive at the same time.
+;             pm.process_idle_timeout   - The number of seconds after which
+;                                         an idle process will be killed.
+; Note: This value is mandatory.
+pm = dynamic
+
+; The number of child processes to be created when pm is set to 'static' and the
+; maximum number of child processes when pm is set to 'dynamic' or 'ondemand'.
+; This value sets the limit on the number of simultaneous requests that will be
+; served. Equivalent to the ApacheMaxClients directive with mpm_prefork.
+; Equivalent to the PHP_FCGI_CHILDREN environment variable in the original PHP
+; CGI. The below defaults are based on a server without much resources. Don't
+; forget to tweak pm.* to fit your needs.
+; Note: Used when pm is set to 'static', 'dynamic' or 'ondemand'
+; Note: This value is mandatory.
+pm.max_children = 30
```

Comment:
Originally the default of 5.

Reply:
This is based on a rough calculation of Journal on its `t2.small` (then rounded down a bit). Might not be too high for some (Drupal?).

Reply:
Maybe parametrise? (Or calculate?)

Reply:
In general this is usually proportional to the number of CPU (cores), so it should be a multiple of ``{% salt['grains.get']('num_cpus') %}``, which is `1` on `t2.small`. Same for other related values like `pm.start_servers` maybe.

Reply:
Advice seems to be available memory ÷ memory required for a request.

Reply:
Memory is definitely an hard limit, `{% salt['grains.get']('mem_total') %}` is `2000` on `journal` servers.

Reply:
`memory_limit` is currently `128M`, so combined that's more than available for Journal. Though I don't believe requests get anywhere near that limit, so could be reduced. Looking to see if can find out how much memory requests use.

Reply:
Think we'd have to add a custom parameter to New Relic then learn how to use Insights.

Reply:
Didn't find a way to do that in New Relic Infrastructure, the metrics being reported are not configurable.

Reply:
Maybe just set `memory_limit` in Pillar, then calculate this as `(mem_total × .8) ÷ memory_limit` (on basis that other services are run elsewhere). Would give 12 for Journal, which is definitely better (`memory_limit` _is_ far too high there, locally it looks like the archive pages use the most memory (not paginated), <40).

Reply:
Article 06416 uses 48MB locally (giant list of authors), could well be the biggest article JSON?

Reply:
`memory_limit` can be tuned for journal but we need a large default to prevent, for example, journal-cms from crashing on half the API calls. The pillar should go into `pillar.elife` so it's easy to override.

Reply:
I'd go the other way, make it 64MB by default then increase where needed.

Reply:
As long as the risky projects have an override from the get-go, `journal-cms` for sure, the others should be fine.

Reply:
To be added here: https://github.com/elifesciences/builder-private/blob/master/pillar/journal-cms.sls#L49